crates/sui-types/src/multisig.rs

```diff
@@ -0,0 +1,316 @@
+                bytes.extend_from_slice(
+                    bcs::to_bytes(self)
+                        .expect("BCS serialization should not fail")
+                        .as_slice(),
+                );
+                Ok(bytes)
+            })
+            .expect("OnceCell invariant violated")
+    }
+}
+
+/// Trait useful to get the bytes reference for [enum GenericSignature].
+impl AsRef<[u8]> for GenericSignature {
+    fn as_ref(&self) -> &[u8] {
+        match self {
+            GenericSignature::MultiSignature(s) => s.as_ref(),
+            GenericSignature::Signature(s) => s.as_ref(),
+        }
+    }
+}
+
+// A macro to implement [trait Serialize] and [trait Deserialize] for [enum GenericSignature] using its bytes representation.
+serde_to_from_bytes!(GenericSignature);
+
+/// The struct that contains signatures and public keys necessary for authenticating a multisig.
+#[serde_as]
+#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
+pub struct MultiSignature {
+    /// The plain signature encoded with signature scheme.
+    sigs: Vec<CompressedSignature>,
+    /// A bitmap that indicates the position of which public key the signature should be authenticated with.
+    #[schemars(with = "Base64")]
+    #[serde_as(as = "SuiBitmap")]
+    bitmap: RoaringBitmap,
+    /// The public key encoded with each public key with its signature scheme used along with the corresponding weight.
+    multi_pk: MultiPublicKey,
+    /// A bytes representation of [struct MultiSignature]. This helps with implementing [trait AsRef<[u8]>].
+    #[serde(skip)]
+    bytes: OnceCell<Vec<u8>>,
+}
+
+/// Necessary trait for [struct SenderSignedData].
+impl PartialEq for MultiSignature {
+    fn eq(&self, other: &Self) -> bool {
+        self.sigs == other.sigs && self.bitmap == other.bitmap && self.multi_pk == other.multi_pk
+    }
+}
+
+/// Necessary trait for [struct SenderSignedData].
+impl Eq for MultiSignature {}
+
+/// Necessary trait for [struct SenderSignedData].
+impl Hash for MultiSignature {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.as_ref().hash(state);
+    }
+}
+
+impl AuthenticatorTrait for MultiSignature {
+    fn verify_secure_generic<T>(
+        &self,
+        value: &IntentMessage<T>,
+        author: SuiAddress,
+    ) -> Result<(), SuiError>
+    where
+        T: Serialize,
+    {
+        if self.multi_pk.pk_map.len() > MAX_SIGNER_IN_MULTISIG {
+            return Err(SuiError::InvalidSignature {
+                error: "Invalid number of public keys".to_string(),
+            });
+        }
+
+        if <SuiAddress as From<MultiPublicKey>>::from(self.multi_pk.clone()) != author {
+            return Err(SuiError::InvalidSignature {
+                error: "Invalid address".to_string(),
+            });
+        }
+        let mut weight_sum = 0;
+        let message = bcs::to_bytes(&value).expect("Message serialization should not fail");
+
+        for (sig, i) in self.sigs.iter().zip(&self.bitmap) {
```

Comment:
Does it assume here that the sigs are sorted according to their public keys?

Reply:
also, we depend here on bitmap to check that we don't have 2 signatures for the same public key, right?

Reply:
yeah sigs are assumed sorted

Reply:
yeah roaring bitmap prevents inserting the same value twice. 

however, it is technically possible if the multisig pubkey itself is constructed with two same pks. i dont know if there is a good reason someone would do that but didnt check to fail for this.